core/src/main/java/nu/yona/server/subscriptions/service/UserService.java

```diff
@@ -288,4 +308,20 @@ public void assertValidEmailAddress(String emailAddress)
 	{
 		UserAssertionService.assertValidEmailAddress(emailAddress);
 	}
+
+	/**
+	 * Locks the user entity with the given ID for update, thus preventing concurrent updates on that user. To prevent from using
+	 * stale data that was retrieved before any previous update completed, the Hibernate session is cleared. This operation fails
+	 * if the session is dirty. Callers should obtain the lock on the user entity prior to any checks or updates. Note that this
+	 * operation is idempotent: if the user was already locked during this session, that user will be returned. The session won't
+	 * be cleared.
+	 *
+	 * @param userId The ID of the user to lock
+	 * @return The locked user entity
+	 */
+	@Transactional
+	public User lockUserForUpdate(UUID userId)
```

Comment:
How does this work when `lockUserForUpdate` is called and an exception occurs? Is the lock released because the transaction failed?

Reply:
Indeed. Once the transaction completes (whether successfully or in error), the lock is released.